mozilla/ichnaea
# Add load_cell_data.py for public cell exports

``ichnaea/scripts/load_cell_data.py`` reads a differential cell export file and creates or updates the related cell records. It also queues the tasks to process the related cell areas (normally done in the submission and query pipelines) and region statistics (normally done on a schedule).

Test process:
1. Download a few differential cell exports from https://location.services.mozilla.com/downloads to the repository.
2. Start celery with ``docker-compose up worker``. This will just run the celery worker, while ``make runcelery`` will also run the scheduler, making it harder to follow what it going on.
3. In a second shell, start a server with ``make run``. http://localhost:8000/stats/regions should show no data.
4. In a third shell, run ``make shell`` and start working with the new code. Import one of the cell exports (start with an older one):

    ```
    # Replace with the name of the export
    ichnaea/scripts/load_cell_data.py MLS-diff-cell-export-YYYY-DD-MMTHH0000.csv.gz
    ```

It should take 30-60 seconds, and look something like this:

```
2019-10-08 15:44:17,490 DEBUG [ichnaea.log.DebugRavenClient] - Configuring Raven for host: None
2019-10-08 15:44:19,558 INFO  [ichnaea.data.public] - Processed 1000 stations
2019-10-08 15:44:21,462 INFO  [ichnaea.data.public] - Processed 2000 stations
2019-10-08 15:44:23,471 INFO  [ichnaea.data.public] - Processed 3000 stations
2019-10-08 15:44:25,463 INFO  [ichnaea.data.public] - Processed 4000 stations
2019-10-08 15:44:26,824 INFO  [ichnaea.data.public] - Processed 1000 station areas
2019-10-08 15:44:27,342 INFO  [ichnaea.data.public] - Processed 5000 stations
2019-10-08 15:44:29,423 INFO  [ichnaea.data.public] - Processed 6000 stations
2019-10-08 15:44:31,513 INFO  [ichnaea.data.public] - Processed 7000 stations
2019-10-08 15:44:33,532 INFO  [ichnaea.data.public] - Processed 8000 stations
2019-10-08 15:44:35,175 INFO  [ichnaea.data.public] - Processed 2000 station areas
2019-10-08 15:44:35,493 INFO  [ichnaea.data.public] - Processed 9000 stations
2019-10-08 15:44:37,570 INFO  [ichnaea.data.public] - Processed 10000 stations
2019-10-08 15:44:39,705 INFO  [ichnaea.data.public] - Processed 11000 stations
2019-10-08 15:44:39,732 INFO  [ichnaea.data.public] - Complete, processed 11008 stations:
2019-10-08 15:44:39,732 INFO  [ichnaea.data.public] -   gsm: 2903 new, 0 updated, 0 already loaded
2019-10-08 15:44:39,732 INFO  [ichnaea.data.public] -   lte: 2310 new, 0 updated, 0 already loaded
2019-10-08 15:44:39,732 INFO  [ichnaea.data.public] -   wcdma: 5795 new, 0 updated, 0 already loaded
2019-10-08 15:44:39,733 INFO  [ichnaea.data.public] -   2537 station areas updated
```

The celery worker will show the ``update_cellarea``, ``load_cellarea``, and ``update_statregion`` tasks being received and completing successfully. The ``update_cellarea`` task processes a chunk of cell area data, and reschedules itself while chunks remain.

http://localhost:8000/stats/regions should show some cell data.

Other things to try:

* Load the same cell export file twice. It should succeed, and show all the stations as already loaded.
* Load an older cell export file. It should succeed, with a mix of new and already loaded stations.
* Load a newer cell export file. It should succeed, with a mix of new, updated, and already loaded stations.

This is part of issue #503, but there's a lot more to that issue, even with the reduced scope.
